packages/notifier/tools/subscribe.js

```diff
@@ -0,0 +1,148 @@
+  Far('AsyncIterable', {
+    [Symbol.asyncIterator]: () => {
+      const it = E(itP)[Symbol.asyncIterator]();
+      return Far('AsyncIterator', {
+        next: async () => E(it).next(),
+      });
+    },
+  });
+
+/**
+ * Asyncronously iterates over the contents of a PublicationList as they appear.
+ * This iteration must drop parts of publication records that are no longer
+ * needed so they can be garbage collected.
+ *
+ * @template T
+ * @param {PublicationList<T>} pubList
+ * @returns {AsyncIterableIterator<T, T>}
+ */
+const makeEachIterator = pubList => {
+  // To understand the implementation, start with
+  // https://web.archive.org/web/20160404122250/http://wiki.ecmascript.org/doku.php?id=strawman:concurrency#infinite_queue
+  return Far('EachIterator', {
+    next: () => {
+      const resultP = E.get(pubList).head;
+      pubList = E.get(pubList).tail;
+      // We expect the tail to be the "cannot read past end" error at the end
+      // of the happy path.
+      // Since we are wrapping that error with eventual send, we sink the
+      // rejection here too so it doesn't become an invalid unhandled rejection
+      // later.
+      void E.when(pubList, sink, sink);
+      return resultP;
+    },
+    [Symbol.asyncIterator]: () => makeEachIterator(pubList),
+  });
+};
+
+/**
+ * Given a local or remote subscriber, returns a local AsyncIterable which
+ * provides "prefix lossy" iterations of the underlying PublicationList.
+ * By "prefix lossy", we mean that you may miss everything published before
+ * you ask the returned iterable for an iterator. But the returned iterator
+ * will enumerate each thing published from that iterator's starting point.
+ *
+ * If the underlying PublicationList is terminated, that terminal value will be
+ * reported losslessly.
+ *
+ * @template T
+ * @param {ERef<EachTopic<T>>} topic
+ */
+export const subscribeEach = topic => {
+  const iterable = Far('EachIterable', {
+    [Symbol.asyncIterator]: () => {
+      const pubList = E(topic).subscribeAfter();
+      return makeEachIterator(pubList);
+    },
+  });
+  return iterable;
+};
+harden(subscribeEach);
+
+/**
+ * @template T
+ * @param {ERef<LatestTopic<T>>} topic
+ * @param {UpdateCount} [localUpdateCount]
+ * @returns {AsyncIterableIterator<T>}
+ */
+const makeLatestIterator = (topic, localUpdateCount) => {
+  let myIterationResultP;
+  return Far('LatestIterator', {
+    [Symbol.asyncIterator]: () => makeLatestIterator(topic, localUpdateCount),
+    next: () => {
+      // In this adaptor, once `next()` is called and returns an
+      // unresolved promise, `myIterationResultP`, and until
+      // `myIterationResultP` is fulfilled with an
+      // iteration result, further `next()` calls will return the same
+      // `myIterationResultP` promise again without asking the notifier
+      // for more updates. If there's already an unanswered ask in the
+      // air, all further asks should just reuse the result of that one.
```

Comment:
This seems... bad. I don't think we ever want _any_ series of successive `next()` calls to observe artificially inserted results, and would rather throw an error than commit to such a pattern.

Reply:
I didn't change this code.  Again, referring to @erights for judgement on this feature.

Reply:
@gibson042 , I do not follow your reasoning here. Given the nature of the lossiness of "latest", this seems like the desired behavior. Could you explain why not? Thanks.

Reply:
Lossy iteration is good, it's "gainy" iteration that I have a problem with. [docs.agoric.com](https://docs.agoric.com/guides/js-programming/notifiers.html#notifierkit-and-subscriptionkit) describes the iteration in terms of subsets where "_Every non-final value in the sampling subset appears in producer order_", which this code violates. It also deviates from standard ECMAScript behavior in a way that will hinder future attempts to accommodate batching. For example, observe how the iterator from an async generator function resolves all results with successive values until the sequence terminates, even if its `next` method is invoked many times before the first result has settled:
```js
const now = () => Math.floor(performance.now());
const gen = async function* (n) {
  while (n--) {
    yield now();
    await new Promise(resolve => setTimeout(resolve, 1000));
  }
  console.log({ iterCloseTime: now() });
};
const iter = gen(3);
Array(5).fill().forEach(async () => {
  const callTime = now();
  const result = await iter.next();
  console.log({ callTime, result, receiveTime: now()});
});
/* logs:
{
  callTime: 31,
  result: { value: 31, done: false },
  receiveTime: 31
}
{
  callTime: 31,
  result: { value: 1033, done: false },
  receiveTime: 1033
}
{
  callTime: 31,
  result: { value: 2035, done: false },
  receiveTime: 2035
}
{ iterCloseTime: 3037 }
{
  callTime: 31,
  result: { value: undefined, done: true },
  receiveTime: 3038
}
{
  callTime: 31,
  result: { value: undefined, done: true },
  receiveTime: 3038
}
*/
```

Reply:
The "gainy" vs subset argument makes sense to me. I buy it. Thanks!

Reply:
PTAL at my new implementation that produces a true subset: https://github.com/Agoric/agoric-sdk/pull/6866#discussion_r1116314783